src/coreclr/jit/flowgraph.cpp

```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+    //
+    // if (comp->m_newToOldLoop[loop->GetIndex()] != nullptr)
+    // {
+    //   printf(" (old: " FMT_LP ")", (unsigned)(comp->m_newToOldLoop[loop->GetIndex()] - comp->optLoopTable));
+    // }
+    //
+    // However, not all callers of FlowGraphNaturalLoops::Find update m_newToOldLoop -- only
+    // Compiler::optFindNewLoops() does that. This dumper should work with any construction of
+    // FlowGraphNaturalLoops.
+
+    printf(" header: " FMT_BB, loop->GetHeader()->bbNum);
+    if (loop->GetParent() != nullptr)
+    {
+        printf(" parent: " FMT_LP, loop->GetParent()->GetIndex());
+    }
+
+    // Dump the set of blocks in the loop. There are three cases:
+    // 1. If there is only one block in the loop, display it.
+    // 2. If the blocks happen to be lexically dense and without non-loop blocks in the range,
+    //    then use a shortcut of `[BBtop .. BBbottom]`.
+    // 3. Otherwise, display the entire list of blocks.
+    //
+    // Lexicality depends on properly renumbered blocks, which we might not have when dumping.
+
+    printf("\n  Members (%u): ", loop->NumLoopBlocks());
+
+    const unsigned numBlocks = loop->NumLoopBlocks();
+    if (numBlocks == 0)
+    {
+        // This should never happen?
+        printf("NONE?");
+    }
+    else if (numBlocks == 1)
+    {
+        // Should be able to assume the only block is the header, but just in case, use the normal visitor.
+        loop->VisitLoopBlocksReversePostOrder([](BasicBlock* block) {
+            printf(FMT_BB " ", block->bbNum);
+            return BasicBlockVisit::Continue;
+        });
+    }
+    else
+    {
+        BasicBlock* lexicalTopBlock     = loop->GetLexicallyTopMostBlock();
+        BasicBlock* lexicalBottomBlock  = loop->GetLexicallyBottomMostBlock();
+        BasicBlock* lexicalEndIteration = lexicalBottomBlock->Next();
+
+        // Add all the loop blocks to a set, then check all the blocks in the identified lexical range.
+        // If there are non-loop blocks found, or if we don't find all the loop blocks in the lexical walk
+        // (meaning the bbNums might not be properly ordered), we fail.
+        loop->VisitLoopBlocks([&](BasicBlock* loopBlock) {
+            BitVecOps::AddElemD(&loopTraits, outputBlocks, loopBlock->bbNewPostorderNum);
+            return BasicBlockVisit::Continue;
+        });
+        bool lexicallyDense = true; // assume the best
+        for (BasicBlock* block = lexicalTopBlock; (block != nullptr) && (block != lexicalEndIteration);
+             block             = block->Next())
+        {
+            if (!loop->ContainsBlock(block))
+            {
+                lexicallyDense = false;
+                break;
+            }
+            BitVecOps::RemoveElemD(&loopTraits, outputBlocks, block->bbNewPostorderNum);
+        }
+        if (!BitVecOps::IsEmpty(&loopTraits, outputBlocks))
+        {
+            lexicallyDense = false;
+        }
```

Comment:
Instead of the bit vector I think we can use the count of loop blocks seen and compare it to `numBlocks`.

Reply:
How hard would it be to use range notation for each consecutive span of set bits?

Reply:
I would just dump things in `bbNum` order as the new loop structures are `bbNum` dependent. If we want something semi-durable (eg is this the "same" loop we saw earlier?) we should perhaps be thinking of dumping the `bbID`.

Reply:
> the new loop structures are bbNum dependent

I don't think that's true: they use a new `bbNewPostorderNum` as the key to the loop member bitset, so `bbNum` ordering is only used for "legacy" lexicality uses.

Reply:
Ah right, that's something Jakob changed from the old synthesis only version....  we can just go with what you have now and see how well it holds up.

Reply:
I went ahead and implemented case (1), above. It works out pretty well most of the time.

When we build loops for the alignment placing phase, the `bbNum` are very unordered w.r.t. `bbNext`, so we end up falling back to just displaying all the loop block individually.